### PR TITLE
feat: support drag drop for dashboard

### PR DESCRIPTION
**Pull request by ugur-vaadin**
## Description

Adds drag drop support for dashboard.

Added API:
- `void setEditable(boolean editable)`: Sets the option to make the dashboard editable
- `boolean isEditable()`: Returns whether the dashboard is editable
- `Registration addItemReorderStartListener(ComponentEventListener listener)`: Adds an item reorder start listener to the dashboard
- `Registration addItemReorderEndListener(ComponentEventListener listener)`: Adds an item reorder end listener to the dashboard

Fixes https://github.com/orgs/vaadin/projects/70/views/1?pane=issue&itemId=78740411

Part of https://github.com/vaadin/platform/issues/6626

## Type of change

Feature

### CODE REVIEW

File: vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java [hunk 1]
```diff
@@ -390,4 +448,151 @@ Collection<Component> getDirectChildren() {
...
+            // Assert section child item node ID
+            if (section.getWidgets().stream().noneMatch(child -> child
+                    .getElement().getNode().getId() == sectionItemNodeId)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void customizeItemReorderEndEvent() {
```
Comment (ugur-vaadin) [inline vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java:573]: I added this back because I encountered the same problem even after using just the `Object` to work around the serialization issue.

Comment (tomivirkki) [inline vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java:573]: This needs to run on attach, otherwise it wouldn't work if you for example detach and reattach the dashboard:
```java
add(new NativeButton("Toggle attached", e -> {
        if (dashboard.getParent().isPresent()) {
                dashboard.removeFromParent();
        } else {
                add(dashboard);
        }
}));
```

File: vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java [hunk 2]
```diff
@@ -390,4 +448,151 @@ Collection<Component> getDirectChildren() {
...
+                                const sectionItemCopy = { nodeid: sectionItems[sectionItemIdx].nodeid };
+                                sectionItemsCopy.push(sectionItemCopy);
+                              }
+                              itemCopy.items = sectionItemsCopy;
+                            }
+                            itemsCopy.push(itemCopy);
+                          }
+                          const flowReorderEvent = new CustomEvent('dashboard-item-reorder-end-flow', {
+                            detail: { items: itemsCopy },
+                          });
```
Comment (tomivirkki) [inline vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java:594]: Could shorten this to:
```js
function mapItems(items) {
  return items.map(({nodeid, items}) => ({
    nodeid,
    ...(items && { items: mapItems(items) })
  }));
}

const flowReorderEvent = new CustomEvent('dashboard-item-reorder-end-flow', {
  detail: { items: mapItems(this.items) }
});
```

File: vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java (hunk 1 above)
Comment (ugur-vaadin) [reply vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java:567]: Fixed the issue with dashboard detach&reattach and added tests.

File: vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java (hunk 2 above)
Comment (ugur-vaadin) [reply vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java]: Updated the code to use this shortened version.

File: vaadin-dashboard-flow-parent/vaadin-dashboard-flow-integration-tests/src/test/java/com/vaadin/flow/component/dashboard/tests/DashboardDragDropIT.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    private void reorderRootLevelItem(int initialIndex, int targetIndex) {
+        executeScript(
+                """
+                        const reorderedItem = arguments[0].items.splice(%d, 1)[0];
+                        arguments[0].items.splice(%d, 0, reorderedItem);
+                        arguments[0].dispatchEvent(new CustomEvent('dashboard-item-reorder-end'));"""
+                        .formatted(initialIndex, targetIndex),
+                dashboardElement);
```
Comment (tomivirkki) [inline vaadin-dashboard-flow-parent/vaadin-dashboard-flow-integration-tests/src/test/java/com/vaadin/flow/component/dashboard/tests/DashboardDragDropIT.java:92]: Could consider using Selenium for DnD. I'm fine with the JS approach also but it feels somewhat hacky.
```java
var initialWidget = dashboardElement.getWidgets().get(initialIndex);
var dragHandle = initialWidget.$("*").withClassName("drag-handle").first();

var targetWidget = dashboardElement.getWidgets().get(targetIndex);

var yOffset = initialWidget.getLocation().getY() < targetWidget.getLocation().getY() ? 10 : -10;
var xOffset = initialWidget.getLocation().getX() < targetWidget.getLocation().getX() ? 10 : -10;

new Actions(driver).clickAndHold(dragHandle)
        .moveToElement(targetWidget, xOffset, yOffset)
        .release(targetWidget)
        .build()
        .perform();
```

Comment (ugur-vaadin) [reply vaadin-dashboard-flow-parent/vaadin-dashboard-flow-integration-tests/src/test/java/com/vaadin/flow/component/dashboard/tests/DashboardDragDropIT.java]: Done. Also added tests for dragging items when editable is toggled.

File: vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java [hunk 4]
```diff
@@ -48,6 +57,8 @@ public class Dashboard extends Component implements HasWidgets {
      */
     public Dashboard() {
         childDetachHandler = getChildDetachHandler();
+        setEditable(true);
```
[the remark this answers is not in the document]
Comment (ugur-vaadin) [reply vaadin-dashboard-flow-parent/vaadin-dashboard-flow/src/main/java/com/vaadin/flow/component/dashboard/Dashboard.java]: Done.

**Review by tomivirkki — APPROVED** [review]
